File: 05_arrays/05_22_four_sum/four_sum.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> fourSum(vector<int> &nums, int target)
{
    sort(nums.begin(), nums.end());
    int size = nums.size();
    vector<vector<int>> result;

    for (int i = 0; i < size - 3; i++)
    {
        if (i > 0 && nums[i] == nums[i - 1])
        {
            continue;
        }
        for (int j = i + 1; j < size - 2; j++)
        {
            if (j > i + 1 && nums[j] == nums[j - 1])
            {
                continue;
            }
            int low = j + 1;
            int high = size - 1;

            while (low < high)
            {
                long long sum = static_cast<long long>(nums[i]) + nums[j] + nums[low] + nums[high];

                if (sum < target)
                {
                    low++;
                }
                else if (sum > target)
                {
                    high--;
                }
                else
                {
                    vector<int> combination = {nums[i], nums[j], nums[low], nums[high]};
                    result.push_back(combination);

                    while (low + 1 < high && nums[low] == nums[low + 1])
                    {
                        low++;
                    }
                    while (high - 1 > low && nums[high] == nums[high - 1])
                    {
                        high--;
                    }

                    low++;
                    high--;
                }
            }
        }
    }

    return result;
}

int main() {
    vector<int> nums1 = {1, 0, -1, 0, -2, 2};
    int target1 = 0;
    vector<vector<int>> result1 = fourSum(nums1, target1);
    cout << "Example 1:\n";
    for (const auto& quadruplet : result1) {
        cout << "[";
        for (int num : quadruplet) {
            cout << num << " ";
        }
        cout << "]\n";
    }

    vector<int> nums2 = {2, 2, 2, 2, 2};
    int target2 = 8;
    vector<vector<int>> result2 = fourSum(nums2, target2);
    cout << "\nExample 2:\n";
    for (const auto& quadruplet : result2) {
        cout << "[";
        for (int num : quadruplet) {
            cout << num << " ";
        }
        cout << "]\n";
    }

    return 0;
}

File: 05_arrays/05_22_four_sum/four_sum.py
def four_sum(nums: list, target: int):
    result = []
    size = len(nums)
    if size < 4:
        return result
    nums.sort()
    for i in range(size - 3):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        for j in range(i + 1, size - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue
            low = j + 1
            high = size - 1
            while low < high:
                sumation = nums[i] + nums[j] + nums[low] + nums[high]
                if sumation < target:
                    low += 1
                elif sumation > target:
                    high -= 1
                else:
                    result.append([nums[i], nums[j], nums[low], nums[high]])
                    while low + 1 < high and nums[low] == nums[low + 1]:
                        low += 1
                    while high - 1 > low and nums[high] == nums[high - 1]:
                        high -= 1
                    low += 1
                    high -= 1
    return result


# Example 1
nums1 = [1, 0, -1, 0, -2, 2]
target1 = 0
result1 = four_sum(nums1, target1)
print("Example 1:")
print(result1)

# Example 2
nums2 = [2, 2, 2, 2, 2]
target2 = 8
result2 = four_sum(nums2, target2)
print("\nExample 2:")
print(result2)